### PR TITLE
London | 26-ITP-Jan | Karla Grajales| Sprint 1 | Data Groups

### PR DESCRIPTION
**Pull request by Grajales-K**
## Learners, PR Template

Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist

In this project, I mastered TDD (Test Driven Development) while implementing various functions. I learned different ways to run tests, such as using npm test and npx jest --watch to run specific tests or all of them at once.

I also learned how to transition from a classic for loop to a for...of loop, which is simpler and more readable. I gained experience using array methods like .filter() and .reduce(), working with Math.max(), using the modulo operator (%), and manipulating arrays with the spread operator (...). This has given me a more confident and solid foundation in TDD.

## Questions

I noticed that without the compare function (a, b) => a - b, the .sort() method treats numbers as strings (so '110' comes before '20'). Is it considered best practice to always provide a compare function even if we think the data is already clean?

Thank you for your time reviewing my code.

### CODE REVIEW

File: Sprint-1/fix/median.js [hunk 1]
```diff
@@ -5,10 +5,45 @@
...
+
+function calculateMedian(list) { 
+  // if is not an array 
+  if(!Array.isArray(list)){
+    return null;
+  }
+
+  //checking only numbers and not null and empty
+  const numsOnly = list.filter(
+    (n) => typeof n === "number" && n !== null && !isNaN(n)
```
Comment (cjyuan) [inline Sprint-1/fix/median.js]: One of the checks is redundant.

File: Sprint-1/fix/median.js [hunk 2]
```diff
@@ -5,10 +5,45 @@
...
+  );
+
+  //if return numbers check if after cleaning the length is same as expected or empty
+  if (numsOnly.length === 0) {
+    return null;
+  }
+
+  // now we change to use the numsOnly otherwise will take string as numbers
+  const sortArr = [...numsOnly].sort((a, b) => a - b);
```
Comment (cjyuan) [inline Sprint-1/fix/median.js]: Why make a copy of `numsOnly` before sorting?

File: Sprint-1/implement/dedupe.js [hunk 3]
```diff
@@ -1 +1,9 @@
-function dedupe() {}
+
+
+function dedupe(val) {
```
Comment (cjyuan) [inline Sprint-1/implement/dedupe.js]: Can you think of a name that could more clearly indicate the parameter is an array of numbers?

File: Sprint-1/implement/dedupe.test.js [hunk 4]
```diff
@@ -16,13 +16,31 @@ E.g. dedupe([1, 2, 1]) returns [1, 2]
...
+test("given an empty array, it returns an empty array", () => {
+  expect(dedupe([])).toEqual([]);
+});
 
 // Given an array with no duplicates
 // When passed to the dedupe function
 // Then it should return a copy of the original array
+test("given an array with no duplicates, it returns a copy of the original array", () => {
+  expect(dedupe(["a", "b", "c"])).toEqual(["a", "b", "c"]);
+});
```
Comment (cjyuan) [inline Sprint-1/implement/dedupe.test.js:33]: This test should fail if the function returns the original array (instead of a **copy** of the original array).

The current test checks only if both the original array and the returned array contain identical elements. 
In order to validate the returned array is a different array, we need an **additional** check.

Can you find out what this additional check is?

File: Sprint-1/implement/max.test.js [hunk 5]
```diff
@@ -5,39 +5,63 @@ In this kata, you will need to implement a function that find the largest numeri
...
+});
+
 
 // Given an array with only non-number values
 // When passed to the max function
 // Then it should return the least surprising value given how it behaves for all other inputs
+test("given an array with only non-number values, it returns -Infinity", () => {
+  expect(findMax(["abc", "def"])).toEqual(-Infinity);
+});
+
```
Comment (cjyuan) [inline Sprint-1/implement/max.test.js:81]: When a string representing a valid numeric literal (for example, `"300"`) is compared to a number, 
JavaScript first converts the string into its numeric equivalent before performing the comparison. 
As a result, the expression `20 < "300"` evaluates to `true`.

To test if the function can correctly ignore non-numeric values,
consider including a string such as `"300"` in the relevant test cases.

File: Sprint-1/implement/max.js [hunk 6]
```diff
@@ -1,4 +1,10 @@
+// 1. filter: leep only elements that are onlyNumbers.
+// 2. spread (...) take the numbers out of the Array
+// 3. Math.max find the largest number and return -infinity if empty.
+
 function findMax(elements) {
+    const onlyNumbers = elements.filter((element) => typeof element === "number")
+    return Math.max(...onlyNumbers);
 }
```
Comment (cjyuan) [inline Sprint-1/implement/max.js:8]: What do you expect from the following function calls (on extreme cases)?
Does your function return the value you expected?

```
findMax([NaN])
findMax([0, NaN, 1])
```

File: Sprint-1/implement/sum.js [hunk 7]
```diff
@@ -1,4 +1,6 @@
 function sum(elements) {
+
+    return elements.reduce((acc, curr) => (typeof curr === "number" ? acc + curr : acc), 0)
 }
```
Comment (cjyuan) [inline Sprint-1/implement/sum.js:7]: What do you expect from the following function calls (on extreme cases)?
Does your function return the value you expected?

```
sum([NaN, 1]);
sum([Infinity, -Infinity]);
```

File: Sprint-1/implement/sum.test.js [hunk 8]
```diff
@@ -13,24 +13,41 @@ const sum = require("./sum.js");
...
+test("Should return the correct sum even when the array contains negative numbers", () => {
+  expect(sum([-1, -2, -3])).toEqual(-6);
+});
 
 // Given an array with decimal/float numbers
 // When passed to the sum function
 // Then it should return the correct total sum
+test("Should return the correct sum even when the array contains decimal numbers", () => {
+  expect(sum([1.4, 1.6, 5.8])).toEqual(8.8);
+});
```
Comment (cjyuan) [inline Sprint-1/implement/sum.test.js:39]: Decimal numbers in most programming languages (including JS) are internally represented in "floating point number" format. Floating point arithmetic is not exact. For example, the result of `46.5678 - 46 === 0.5678` is false because `46.5678 - 46` only yield a value that is very close to `0.5678`. Even changing the order in which the program add/subtract numbers can yield different values.

So the following could happen
```javascript
  expect( 1.2 + 0.6 + 0.005 ).toEqual( 1.805 );                // This fail
  expect( 1.2 + 0.6 + 0.005 ).toEqual( 1.8049999999999997 );   // This pass
  expect( 0.005 + 0.6 + 1.2 ).toEqual( 1.8049999999999997 );   // This fail

  console.log(1.2 + 0.6 + 0.005 == 1.805);  // false
  console.log(1.2 + 0.6 + 0.005 == 0.005 + 0.6 + 1.2); // false
```

**Can you find a more appropriate way to test a value (that involves decimal number calculations) for equality?**

Suggestion: Look up 
- Checking equality in floating point arithmetic in JavaScript 
- Checking equality in floating point arithmetic with Jest

File: Sprint-1/refactor/includes.js [hunk 9]
```diff
@@ -1,12 +1,11 @@
...
-  for (let index = 0; index < list.length; index++) {
-    const element = list[index];
-    if (element === target) {
+function includes(list, target){
+  for (const element of list){
+    if (element === target){
       return true;
     }
-  }
+  }
```
Comment (cjyuan) [inline Sprint-1/refactor/includes.js]: Spacing is a bit inconsistently.

Have you installed the **prettier** VSCode extension and enabled "Format on save/paste" on VSCode,
as recommended in https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/blob/main/readme.md
?

**Comment by cjyuan** [comment]
> I noticed that without the compare function (a, b) => a - b, the .sort() method treats numbers as strings (so '110' comes before '20'). Is it considered best practice to always provide a compare function even if we think the data is already clean?

Have you read the documentation of `Array.prototype.sort()` on MDN to find out how the function works?
I think AI can also explain why.

I am not sure what you meant by "data is already clean".

File: Sprint-1/fix/median.js (hunk 1 above)
Comment (Grajales-K) [reply Sprint-1/fix/median.js]: thanks @cjyuan Removed n !== null because typeof n === 'number' already excludes nulls. Since it's an && check, if the first part is false, JavaScript short-circuits and doesn't evaluate the rest

File: Sprint-1/fix/median.js (hunk 2 above)
Comment (Grajales-K) [reply Sprint-1/fix/median.js]: @cjyuan, I removed the spread operator [...] because .filter() already returns a new array. Sorting numsOnly directly is safe and more efficient since it doesn't mutate the original input.

File: Sprint-1/implement/dedupe.js (hunk 3 above)
Comment (Grajales-K) [reply Sprint-1/implement/dedupe.js]: Hi @cjyuan, I've renamed the parameter to items since the function receives an array of different types, including both strings and numbers.

File: Sprint-1/implement/dedupe.test.js (hunk 4 above)
Comment (Grajales-K) [reply Sprint-1/implement/dedupe.test.js:33]: Hi @cjyuan, I've updated the test to use a variable for the input. This allows me to use not.toBe() to prove that the function returns a new reference (a copy) and not the original array, even though the content is identical.

File: Sprint-1/implement/max.test.js (hunk 5 above)
Comment (Grajales-K) [reply Sprint-1/implement/max.test.js:81]: Hi @cjyuan, I've added the tests with numeric strings (like "300"). My function ignores them because it only accepts actual numbers, which helps avoid any confusion with type coercion.

File: Sprint-1/implement/max.js (hunk 6 above)
Comment (Grajales-K) [reply Sprint-1/implement/max.js:8]: Got it! I added the !Number.isNaN check to the filter and included new test cases for NaN. Now the function correctly ignores them and returns the expected values, just like the median implementation.

File: Sprint-1/implement/sum.js (hunk 7 above)
Comment (Grajales-K) [reply Sprint-1/implement/sum.js:7]: @cjyuan 
The original function returned NaN because it didn't filter out NaN or Infinity. I've updated it to ignore these cases using !Number.isNaN, ensuring that sum([NaN, 1]) returns 1, and an array of only invalid values returns -Infinity for consistency.

File: Sprint-1/implement/sum.test.js (hunk 8 above)
Comment (Grajales-K) [reply Sprint-1/implement/sum.test.js:39]: @cjyuan  I see it.



```

test("Should return the correct sum even when the array contains decimal numbers", () => {

expect(sum([0.1, 0.2, 0.005])).toEqual(1.805);

});

    Expected: 1.805

    Received: 0.30500000000000005
```
```

test("Should return the correct sum even when the array contains decimal numbers", () => {

  expect(sum([0.1, 0.2, 0.005])).toBeCloseTo(0.305);

}); 


  ✓ pass
  
```
  
  I've updated the decimal test to use .toBeCloseTo(). This correctly handles the floating-point precision error I encountered (0.30500000000000005) without needing to convert the function output to a string.

**Comment by Grajales-K** [comment]
@cjyuan Thank you for your review and am sorry for the delay! I was very sick this weekend with migraine.

I've updated the tests to handle floating-point precision using toBeCloseTo(). I also added edge cases for NaN and Infinity using a data-driven approach to ensure the function is robust

Thank you.

File: Sprint-1/implement/dedupe.test.js [hunk 10]
```diff
@@ -16,13 +16,35 @@ E.g. dedupe([1, 2, 1]) returns [1, 2]
...
 // Given an array with no duplicates
 // When passed to the dedupe function
 // Then it should return a copy of the original array
+test("given an array with no duplicates, it returns a copy of the original array", () => {
+  const input = ["a", "b", "c"];
+  const result = dedupe(input);
+
+  expect(result).toEqual(input);
+  expect(result).not.toBe(input);
+});
```
Comment (cjyuan) [inline Sprint-1/implement/dedupe.test.js:33]: There is a chance that, even though `result` has incorrect elements (for example, `[]`), 
the two tests could still pass. Can you figure out why, and then fix the tests accordingly?

File: Sprint-1/implement/max.test.js (hunk 5 above)
Comment (cjyuan) [reply Sprint-1/implement/max.test.js:81]: I don't see this test updated in this file.

In practice, we write tests not only to verify our current implementation, but also to ensure that future changes do not alter the function's expected behavior.

Currently the following function could also pass this test:
```
function findMax(elements) {
    const onlyNumbers = elements.map(Number).filter((el)=>!Number.isNaN(el));
    return Math.max(...onlyNumbers);
}
```

File: Sprint-1/implement/dedupe.test.js (hunk 10 above)
Comment (Grajales-K) [reply Sprint-1/implement/dedupe.test.js:33]: @cjyuan, I have refactored tests to avoid false positives. Instead of comparing the result against the input variable (which could be mutated), I'm now using explicit literal values and verifying that the original array remains unchanged

File: Sprint-1/implement/max.test.js (hunk 5 above)
Comment (Grajales-K) [reply Sprint-1/implement/max.test.js:81]: Hi @cjyuan, my implementation was already using typeof to strictly filter for numbers, the test suite wasn't explicitly challenging that logic. I've now added a test case with a numeric string ('300') to ensure that any future refactors using loose type coercion (like Number()) will fail. Thanks for this

**Comment by cjyuan** [comment]
Changes look good. Well done.

**Comment by illicitonion** [comment]
Closing PR because the January ITP run has finished. Feel free to re-open if you're still working on it.
